src/common/src/array/data_chunk_iter.rs

```diff
@@ -158,30 +166,35 @@ impl PartialEq for RowRef<'_> {
 }
 impl Eq for RowRef<'_> {}
 
+impl Hash for RowRef<'_> {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        let len = self.columns.len();
+        for i in 0..len {
+            self.datum_at(i).hash(state);
+        }
+    }
+}
+
 impl Row for RowRef<'_> {
     fn datum_at(&self, index: usize) -> DatumRef<'_> {
-        debug_assert!(self.idx < self.chunk.capacity());
         // for `RowRef`, the index is always in bound.
-        unsafe { self.chunk.columns()[index].value_at_unchecked(self.idx) }
+        unsafe { self.columns[index].value_at_unchecked(self.idx) }
```

Comment:
The `unsafe` here becomes unsound as we now check `idx` in bound nowhere for `with_columns`. I'd suggest also adding an assertion in `with_columns` (should be an `assert` instead of `debug_assertion`, should also apply to `new` 🤔)